directord/tests/test_drivers_grpcd.py

```diff
@@ -0,0 +1,846 @@
+
+        mock_client_inst.assert_called_once_with(
+            self.driver.log,
+            "127.0.0.1",
+            5558,
+            False,
+            True,
+            "/etc/pki/ca-trust/source/anchors/cm-local-ca.pem",
+            "/etc/directord/grpc/ssl/directord.crt",
+            "/etc/directord/grpc/ssl/directord.key",
+        )
+
+    def test_backend_recv(self):
+        """Test backend recv."""
+        mock_backend_check = mock.MagicMock()
+        mock_backend_check.side_effect = [False, True]
+        mock_data = mock.Mock()
+        mock_data.identity = "identity"
+        mock_data.msg_id = "msg_id"
+        mock_data.control = "control"
+        mock_data.command = "command"
+        mock_data.data = "{}"
+        mock_data.info = "info"
+        mock_data.stdout = "stdout"
+        mock_data.stderr = "stderr"
+
+        self.driver._client.get_message.return_value = ("target", mock_data)
+
+        self.driver.backend_check = mock_backend_check
+
+        data = self.driver.backend_recv()
+
+        self.assertEqual(
+            data,
+            ["msg_id", "control", "command", "{}", "info", "stderr", "stdout"],
+        )
+
+    def test_job_recv(self):
+        """Test job recv."""
+        mock_job_check = mock.MagicMock()
+        mock_job_check.side_effect = [False, True]
+        mock_data = mock.Mock()
+        mock_data.identity = "identity"
+        mock_data.msg_id = "msg_id"
+        mock_data.control = "control"
+        mock_data.command = "command"
+        mock_data.data = "{}"
+        mock_data.info = "info"
+        mock_data.stdout = "stdout"
+        mock_data.stderr = "stderr"
+
+        self.driver._client.get_job.return_value = ("target", mock_data)
+
+        self.driver.job_check = mock_job_check
+
+        data = self.driver.job_recv()
+
+        self.assertEqual(
+            data,
+            [
+                "msg_id",
+                "control",
+                "command",
+                "{}",
+                "info",
+                "stderr",
+                "stdout",
+            ],
+        )
+
+    @mock.patch("directord.utils.get_uuid", return_value="uuid")
+    def test_hearbeat_send(self, mock_uuid):
+        mock_job_send = mock.MagicMock()
+        self.driver.job_send = mock_job_send
+        self.driver.machine_id = "machine_id"
+        self.driver.heartbeat_send(1, 1, 1, "grpcd")
+        mock_job_send.assert_called_once_with(
+            target="DIRECTORD_SERVER",
+            identity="test-node",
+            control="\x05",
+            msg_id="uuid",
+            data=(
+                '{"job_id": "uuid", "version": 1, "host_uptime": 1, '
+                '"agent_uptime": 1, "machine_id": "machine_id", '
+                '"driver": "grpcd"}'
+            ),
+        )
+
+
+class TestGrpcQueueBase(unittest.TestCase):
+    """Test queue base."""
+
+    def setUp(self):
+        grpcd.QueueBase._instance = None
```

Comment:
I suppose that normally, just the instance() call should be enough for a proper initialization?

Reply:
Normally yes, but because it'll be shared across the tests you have to none the instance so you get a new one for each test